### PR TITLE
DXD staking and influence contracts

### PR DESCRIPTION
**Pull request by fnanni-0**
### Features

DXDStake.sol

- [x] Maximum commitment time defined by governance.
- [x] ERC20SnapshotUpgradeable & OwnableUpgradeable.
- [x] Non transferable.
- [x] Mints/burns influence equivalence for each DXD staked/withdrawn.
- [x] A stake commitment can be increased before completion.
- [x] Early withdrawals with penalization.

The contract is very inefficient. Probably around ~80k gas per stake can be saved by packing storage variables together and simplifying how snapshotting works.

DXDInfluence.sol

- [x] Influence formula consists of a linear and an exponential element: `stake.a.tc + stake.b.tc^k`, where `tc` is the time commitment of the given stake amount, `a` and `b` are mutable constants defined by governance and `k` is an immutable constant. `a`,  `b` and `k` are rational numbers.
- [x] Is owned by DXDStake, which has permission to mint and burn influence.
- [x] It's not ERC20, but it exposes `balanceOf()`, `balanceOfAt()`, `totalSupply()`, `totalSupplyAt()` and `decimals()`.
- [x] Notifies the Voting Power Token of stake changes.

### More about the influence formula and its implementation

The idea behind using a complex formula to calculate the influence a DXD staker should have on the DAO governance, is to give more decision power to holders who commit to invest in DXDao in the long term. Someone who will hold DXD for 3 years should probably have more voting power than someone who just got DXD and might sell soon. 

However, using a linear function might be either too unfair to short term holders or to long term holders. Should a 4-years holder have double the influence of a 2-years holder? With the formula proposed here, the DAO can have a lot of flexibility regarding how much governance influence someone has for each additional second they commit to stake DXD.

#### The formula

Let's define influence as `i(t) = a.t + b.t^k`, where a, b and k are rational numbers and 0 < t < t_max. Depending on how a, b, k and t_max are defined, `i(t)` could have one of the following properties:
1. convex, monotonically non-decreasing
1. concave, monotonically non-decreasing
1. convex, monotonically decreasing
1. concave, monotonically decreasing
1. linear (b = 0)

Additionally, `i(t)` could be always positive, always negative or a combination of both along 0 < t < t_max. 

Probably, a, b, k should be defined such that the influence is a convex or concave, monotonically non-decreasing function (i.e. the longer you stake, the more influence you get). Play around with the formula [here](https://www.wolframalpha.com/input?i=-0.00000001+*+x+%2B+0.0005+*+x%5E0.5+from+x+%3D+0+to+x+%3D+126144000).

#### The implementation

`DXDInfluence.sol` creates a snapshot for every DXD stake/unstake. This means that it should be possible to get the influence of each account at any snapshot while allowing governance modifications of `a` and `b`. Let's redefine `i(t)` as:

`influence = sum(stake.a.tc + stake.b.tc^k)` over all stakes for a given account, where tc: time commitment of each stake.

This can be rearranged as:
`influence = a * sum(stake.tc) + b * sum(stake.tc^k)`

From this we can conclude that `k` has to be immutable, because otherwise an immense amount of computation would be needed to recalculate user's balances after modifying `k`. On the other hand, `a` and `b` can be modified at any time by governance, because we can simply store `sum(stake.tc)` and `sum(stake.tc^k)` and later calculate the final influence value on the fly.

### CODE REVIEW

File: contracts/dao/DXDInfluence.sol [hunk 1]
```diff
@@ -14,94 +18,192 @@ interface VotingPower {
...
-    mapping(uint256 => uint256) public snapshotTimes; // snapshotTimes[snapshotId]
-    mapping(uint256 => uint256) public totalSupplies; // nonRegisteredTotalSupplies[snapshotId]
 
-    /// @notice Error when trying to transfer influence
-    error Influence__NoTransfer();
+    SD59x18 public linearFactor;
+    SD59x18 public exponentialFactor;
+    UD60x18 public exponent; // Must be immutable
+
+    mapping(address => mapping(uint256 => CumulativeStake)) public cumulativeStakes;
```
Comment (AugustoL) [inline contracts/dao/DXDInfluence.sol:36]: This are snapshots, right? Maybe lets add snapshot to the name and comment what is indexing.

File: contracts/dao/DXDInfluence.sol [hunk 2]
```diff
@@ -14,94 +18,192 @@ interface VotingPower {
...
+        return getInfluence(cumulativeStake);
+    }
 
-        uint256 snapshotTime = snapshotTimes[snapshotId];
-        uint256 index = stakeTimes[account].findUpperBound(snapshotTime);
+    /**
+     * @dev Calculates influence for the given cumulative stake data point.
+     * @param _cumulativeStake Accumulated stake information on a specific snapshot.
+     */
+    function getInfluence(CumulativeStake storage _cumulativeStake) internal view returns (uint256) {
```
Comment (AugustoL) [inline contracts/dao/DXDInfluence.sol:191]: maybe lets change the name to calcultaeInfluence ?

File: contracts/dao/DataSnapshot.sol [hunk 3]
```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/utils/ArraysUpgradeable.sol";
+
+/**
+ * @dev Inspired in ERC20Snapshot from openzeppelin
+ **/
+
+contract DataSnapshot {
```
Comment (AugustoL) [inline contracts/dao/DataSnapshot.sol:10]: This snapshot is used only for accounts, so maybe a better name would be AccountSnapshot?

File: contracts/dao/DXDStake.sol [hunk 4]
```diff
@@ -13,8 +13,25 @@ import "./DXDInfluence.sol";
 contract DXDStake is OwnableUpgradeable, ERC20SnapshotUpgradeable {
     using SafeERC20Upgradeable for IERC20Upgradeable;
 
+    struct StakeCommitment {
+        uint256 commitmentEnd;
+        uint256 timeCommitment;
+        uint256 stake;
+    }
+
+    uint256 public constant DIVISOR = 10_000;
```
Comment (AugustoL) [inline contracts/dao/DXDStake.sol:22]: This is just used for early withdrawal, so maybe use EARLY_WITHDRAWAL_DIVISOR ?

File: contracts/dao/DXDStake.sol [hunk 5]
```diff
@@ -34,9 +52,49 @@ contract DXDStake is OwnableUpgradeable, ERC20SnapshotUpgradeable {
...
+    function disableEarlyWithdrawal() external onlyOwner {
+        earlyWithdrawalsEnabled = false;
+    }
+
+    /**
+     * @dev Enables early withdrawals of stake commitments that have not finalized yet.
+     * @param _penalty Percentage, expressed in basis points, that will be taken from the stake as penalty.
+     * @param _recipient Recipient of the penalty.
+     */
+    function enableEarlyWithdrawal(uint256 _penalty, address _recipient) external onlyOwner {
```
Comment (AugustoL) [inline contracts/dao/DXDStake.sol:80]: What about also adding a minimum time needed at least to passed for a n early withdrawal to be executed?

**Review by AugustoL — COMMENTED** [review]
First review, I will take a closer look again later, but looks great, awesome job

**Review by MiltonTulli — APPROVED** [review]

**Comment by fnanni-0** [comment]
Merging this even though the review is ongoing and some changes will probably be needed. We can continue discussing and iterating in the main PR that will include #304.

**Comment by MiltonTulli** [comment]
> Merging this even though the review is ongoing and some changes will probably be needed. We can continue discussing and iterating in the main PR that will include #304.

Just for reference, main pr => https://github.com/DXgovernance/dxdao-contracts/pull/309

**Pull request merged**